src/InputAreaWidget.ts

```diff
@@ -0,0 +1,188 @@
+// Copyright (c) Jupyter Development Team.
+// Distributed under the terms of the Modified BSD License.
+
+'use strict';
+
+import * as CodeMirror from 'codemirror';
+import 'codemirror/lib/codemirror.css';
+import 'codemirror/mode/meta';
+
+import {
+  Message
+} from 'phosphor-messaging';
+
+import {
+  ResizeMessage, Widget
+} from 'phosphor-widget';
+
+import {
+  ISignal, Signal
+} from 'phosphor-signaling';
+
+import {
+  IInputAreaViewModel, ITextEditorViewModel, IChangedArgs
+} from './InputAreaViewModel';
+
+import * as dmp from 'diff-match-patch';
+
+
+let diffMatchPatch = new dmp.diff_match_patch()
+
+/**
+ * A widget which hosts a CodeMirror editor.
+ */
+export
+class CodeMirrorWidget extends Widget {
+
+  /**
+   * Construct a CodeMirror widget
+   */
+  constructor(model: ITextEditorViewModel) {
+    super();
+    this.addClass('CodeMirrorWidget');
+    this._model = model;
+    this._editor = CodeMirror(this.node);
+    this.updateMimetype(model.mimetype);
+    this.updateLineNumbers(model.lineNumbers);
+    this.updateFixedHeight(model.fixedHeight);
+    this.updateText(model.text);
+
+    this._editor.on('change', (instance, change) => {
+      this._model.text = this._editor.getDoc().getValue();
+    });
+    model.stateChanged.connect(this._modelUpdate, this);
+  }
+
+  /**
+   * Update the fixedHeight
+   */
+  updateFixedHeight(fixedHeight: boolean) {
```

Comment:
They are essentially message handlers for model change events.  They aren't state, so much as they update state of the DOM.  They can be easily overridden by subclasses.  Is that enough reasons?  What do you think?